### ISSUE TITLE
[c-sharp] plugin may produce invalid C# code

### ISSUE THREAD

**Issue by cklam2** [dotansimha/graphql-code-generator#4248]
**Describe the bug**
Some field names in the schema may contain [reserved keywords](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/) in C#. When generating code from such schema it will result in invalid C# code. Example reserved keywords are `do`, `in`, `as`, `int`, `protected`, etc.

```C#
// Invalid C# code
public enum Visibility {
  private,
  public
}
```

**To Reproduce**

1. My GraphQL schema:

```graphql
schema {
  query: Query
}

enum Visibility {
  private,
  public
}

type Query {
  int: Int
  fixed: Boolean
}
```

2. My `codegen.yml` config file:

```yml
generates:
  src/main/c-sharp/my-org/my-app/Types.cs:
    plugins:
      - c-sharp
```

**Expected behavior**
Reserved keywords should be prefixed with an `@`, eg `@int`. From above sample it would be 
```C#
public enum Visibility {
  @private,
  @public
}
```

Alternative is using upper case and use JsonProperty to keep the original name. May require more work. And may not be possible everywhere.
```C#
public enum Visibility {
  [JsonProperty("private")]
  Private,
  [JsonProperty("public")]
  Public
}
```

Note, only keywords cannot be used. Seems that the [contextual keywords](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/#contextual-keywords) can be used.

**Environment:**
- Can use the current online [GraphQL code generator](https://graphql-code-generator.com/#live-demo). In the dropdown choose C# Schema Types.

**Comment by cklam2** [comment]
Solved with #4250

**Comment by cklam2** [comment]
Mmm, exploring the typescript code and it's set up differently than in C# plugin (cleaner). For this case maybe I should not introduce that `getSafeFieldName` function and instead create a converter similar to the upper case, pascal case, etc so it only uses convertName feature to transform value names. 
Did not yet figure out how to do chaining/pipe in converter. Any thoughts?

**Comment by cklam2** [comment]
Updated the code and rebased onto master. Also took the chance to clean up the unit test so format looks cleaner (new unit tests for this bug solution are added).
@dotansimha code is ready for review/merge, thanks.

**Comment by dotansimha** [comment]
Fixed in `v1.16.0`
